File: Project-source/forge-verifier/go/binding.go
// Soft binding fingerprint — see spec/binding-fingerprint.md for the canonical
// string + hash rules. Hard binding (TPM/SE) is platform-specific and left to callers.

package forgeverifier

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net"
	"os"
	"runtime"
	"strings"
)

// ComputeSoftFingerprint builds `mac|hostname|cpu` per spec, SHA-256 → lowercase hex.
//
// Components that cannot be obtained (rare on supported platforms) substitute the
// literal "unknown" — never silently dropped, otherwise the hash becomes unstable.
func ComputeSoftFingerprint() (string, error) {
	mac, err := primaryMAC()
	if err != nil {
		mac = "unknown"
	}
	host, err := os.Hostname()
	if err != nil || host == "" {
		host = "unknown"
	}
	cpu := fmt.Sprintf("%s:%s:%d",
		safeOrUnknown(runtime.GOARCH),
		safeOrUnknown(runtime.GOOS),
		runtime.NumCPU(), // best-effort; CPU vendor/family/model needs cgo on darwin
	)

	canonical := fmt.Sprintf("%s|%s|%s",
		strings.ToLower(mac),
		strings.ToLower(host),
		cpu,
	)
	sum := sha256.Sum256([]byte(canonical))
	return hex.EncodeToString(sum[:]), nil
}

func primaryMAC() (string, error) {
	ifaces, err := net.Interfaces()
	if err != nil {
		return "", err
	}
	for _, ifc := range ifaces {
		if ifc.Flags&net.FlagLoopback != 0 || ifc.Flags&net.FlagUp == 0 {
			continue
		}
		// Skip virtual / docker / vmware interfaces — heuristic by name
		name := strings.ToLower(ifc.Name)
		if strings.Contains(name, "docker") || strings.Contains(name, "veth") ||
			strings.Contains(name, "vmnet") || strings.Contains(name, "br-") {
			continue
		}
		if len(ifc.HardwareAddr) >= 6 {
			return ifc.HardwareAddr.String(), nil
		}
	}
	return "", fmt.Errorf("no usable physical NIC")
}

func safeOrUnknown(s string) string {
	if s == "" {
		return "unknown"
	}
	return s
}

File: Project-source/forge-verifier/rust/src/binding.rs
//! Soft fingerprint — best-effort host identifier for `binding=fingerprint-soft`.
//!
//! Canonical string: `mac|hostname|cpu` (lowercased), SHA-256 → lowercase hex.
//! Matches the Go / Python / Java / C# SDKs byte-for-byte.

use sha2::{Digest, Sha256};

/// Compute the soft fingerprint for the current host.
///
/// Returns `(hex, canonical)` — the second value is the pre-hash string,
/// useful for debugging when two hosts hash to different values.
pub fn compute_soft_fingerprint() -> String {
    let mac = safe(&primary_mac());
    let hostname = safe(
        &hostname::get()
            .ok()
            .and_then(|s| s.into_string().ok())
            .unwrap_or_default(),
    );
    let cpu = safe(&cpu_description());

    let canonical = format!("{mac}|{hostname}|{cpu}").to_lowercase();
    let mut hasher = Sha256::new();
    hasher.update(canonical.as_bytes());
    hex::encode(hasher.finalize())
}

fn primary_mac() -> String {
    match mac_address::get_mac_address() {
        Ok(Some(addr)) => addr.to_string().to_lowercase(),
        _ => String::new(),
    }
}

fn cpu_description() -> String {
    format!(
        "{}:{}:{}",
        std::env::consts::ARCH,
        std::env::consts::OS,
        num_cpus::get()
    )
}

fn safe(s: &str) -> String {
    if s.is_empty() {
        "unknown".to_string()
    } else {
        s.to_string()
    }
}
